language/vm/vm_runtime/src/native_functions/signature.rs

```diff
@@ -0,0 +1,286 @@
+const INVALID_PUBLIC_KEY_SIZE_FAILURE: u64 = DEFAULT_ERROR_CODE + 9;
+
+pub fn native_ed25519_signature_verification(mut arguments: VecDeque<Local>) -> NativeReturnStatus {
+    if arguments.len() != 3 {
+        return NativeReturnStatus::InvalidArguments;
+    }
+    let msg = pop_arg!(arguments, ByteArray);
+    let pubkey = pop_arg!(arguments, ByteArray);
+    let signature = pop_arg!(arguments, ByteArray);
+
+    let cost = ED25519_COST * msg.len() as u64;
+
+    let sig = match ed25519::Ed25519Signature::try_from(signature.as_bytes()) {
+        Ok(sig) => sig,
+        Err(_) => {
+            return NativeReturnStatus::Aborted {
+                cost,
+                error_code: DEFAULT_ERROR_CODE,
+            }
+        }
+    };
+    let pk = match ed25519::Ed25519PublicKey::try_from(pubkey.as_bytes()) {
+        Ok(pk) => pk,
+        Err(_) => {
+            return NativeReturnStatus::Aborted {
+                cost,
+                error_code: DEFAULT_ERROR_CODE,
+            }
+        }
+    };
+
+    let bool_value = match sig.verify_arbitrary_msg(msg.as_bytes(), &pk) {
+        Ok(()) => true,
+        Err(_) => false,
+    };
+    let return_values = vec![Local::bool(bool_value)];
+    NativeReturnStatus::Success {
+        cost,
+        return_values,
+    }
+}
+
+/// Batch verify a collection of signatures using a bitmap for matching signatures to keys.
+pub fn native_ed25519_threshold_signature_verification(
+    mut arguments: VecDeque<Local>,
+) -> NativeReturnStatus {
+    if arguments.len() != 4 {
+        return NativeReturnStatus::InvalidArguments;
+    }
+    let message = pop_arg!(arguments, ByteArray);
+    let public_keys = pop_arg!(arguments, ByteArray);
+    let signatures = pop_arg!(arguments, ByteArray);
+    let bitmap = pop_arg!(arguments, ByteArray);
+
+    let num_of_sigs = match ed25519_threshold_signature_verification(
+        &bitmap,
+        &signatures,
+        &public_keys,
+        &message,
+        BATCH_ED25519_COST,
+    ) {
+        Ok(num_of_sigs) => num_of_sigs,
+        Err(e) => return e,
+    };
+
+    let cost = ed25519_threshold_signature_verification_cost(num_of_sigs, message.len());
+    let return_values = vec![Local::u64(num_of_sigs)];
+    NativeReturnStatus::Success {
+        cost,
+        return_values,
+    }
+}
+
+fn ed25519_threshold_signature_verification_cost(num_of_sigs: u64, message_len: usize) -> u64 {
+    BATCH_ED25519_COST * num_of_sigs * message_len as u64
+}
+
+fn ed25519_threshold_signature_verification(
+    bitmap: &ByteArray,
+    signatures: &ByteArray,
+    public_keys: &ByteArray,
+    message: &ByteArray,
+    abort_cost: u64,
+) -> std::result::Result<u64, NativeReturnStatus> {
+    let bitvec = BitVec::from_bytes(bitmap.as_bytes());
+
+    let num_of_sigs = sanity_check(&bitvec, &signatures, &public_keys, abort_cost)?;
+    let abort_cost = ed25519_threshold_signature_verification_cost(num_of_sigs, message.len());
```

Comment:
So, if the sanity check fails, we pay the minimum `BATCH_ED25519_COST`, but if the error occurs on verification, independently on where the error occurred, we pay the maximum depending on the num of submitted signatures. Is this the expected behavior?

Reply:
Good catch! I was actually hoping you would notice this ;) 

I was unsure how to gas cost the incremental nature of this computation, and I was hoping you could help work this out. 

For now though, I was just going to take the max gas amount here, but we can fine tune it later.

Reply:
Actually, I'm not against this approach, our policy might eventually be to fully penalize faulty transactions or we might come up with a fixed cost, because this will never go to the chain anyway and  only one validator will actually do the calculations, then discard it. Agreed that we should fine tune when we finalize our gas policy.